### PR TITLE
added profiling ability

### PR DESCRIPTION
**Pull request by pawels-optimizely**

### CODE REVIEW

File: examples/main.go [hunk 1]
```diff
@@ -80,3 +177,26 @@ func main() {
...
 }
+
+var RunMemProfile = "false"
+var RunCPUProfile = "false"
+
+func main() {
+
+	if RunMemProfile == "true" || RunCPUProfile == "true" {
+
+		const RUN_NUMBER = 6
```
Comment (mikecdavis) [inline examples/main.go]: Why 6? Not very "stressful" :)

Comment (pawels-optimizely) [reply examples/main.go]: a minimum number to see what's going on, increasing this does not have any substantial impact on profiling, but I can set it to something bigger

**Comment by pawels-optimizely** [comment]
The purpose of this exercise is to benchmark the same logic as in test app. Here everything goes through the same flow. Our unit tests have a lot of mocking done so I am not sure if it will be worthwhile to benchmark everything there, unless we rewrite some tests.

**Comment by mikeng13** [comment]
I agree with @mikecdavis on not overloading examples, especially the `main.go` file. Another way around this is to add the stress test as its own file or maybe even go with `main_benchmark.go` and run it as a proper benchmark test like mike suggested

File: examples/main_profile_feature.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	"github.com/optimizely/go-sdk/optimizely/entities"
+	"github.com/optimizely/go-sdk/optimizely/notification"
+
+	"github.com/pkg/profile"
+)
+
+// stressTest has everything that test app has. it is used to run profile
+func stressTest() {
+	/*
+		For the test app, the biggest json file is used with 100 entities.
```
Comment (mikecdavis) [inline examples/main_profile_feature.go]: Similar comment about the test app not relevant to the OS community.

File: examples/README.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# Optimizely Go SDK Examples
+
+### Profiling
+Prerequisite: `go get github.com/pkg/profile`
+
+* CPU profile. Execute: `go build -ldflags "-X main.RunCPUProfile=true" main_profile_feature.go && ./main_profile_feature`. It will create cpu.pprof file in your current directory. Then run: `go tool pprof -http=:8080 cpu.pprof` and profile cpu usage using web browser.
```
Comment (mikecdavis) [inline examples/README.md]: should we add a script for this?

File: examples/main_profile_feature.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+		DATAFILES_DIR has to be set to point to the path where 100_entities.json is located.
+	*/
+
+	var datafileDir = path.Join(os.Getenv("DATAFILES_DIR"), "100_entities.json")
+
+	datafile, err := ioutil.ReadFile(datafileDir)
+	if err != nil {
+		log.Fatal(err)
```
Comment (mikecdavis) [inline examples/main_profile_feature.go]: should we panic to bail out of the prof?

File: examples/main_profile_feature.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+	// Creates a default, canceleable context
+	notificationCenter := notification.NewNotificationCenter()
+	decisionService := decision.NewCompositeService(notificationCenter)
+
+	clientOptions := client.Options{
+		DecisionService: decisionService,
+	}
+	clientApp, err := optlyClient.ClientWithOptions(clientOptions)
+	if err != nil {
+		log.Fatal(err)
```
Comment (mikecdavis) [inline examples/main_profile_feature.go]: similar comment above

File: optimizely/config/datafileprojectconfig/json_parser_test.go [hunk 6]
```diff
@@ -59,3 +62,18 @@ func TestParseDatafilePasses(t *testing.T) {
 
 	assert.Equal(t, expectedDatafile, parsedDatafile)
 }
+
+func BenchmarkParseDatafilePasses(b *testing.B) {
+	for n := 0; n < b.N; n++ {
+		datafile, err := ioutil.ReadFile("test/100_entities.json")
```
[the remark this answers is not in the document]
Comment (pawels-optimizely) [reply optimizely/config/datafileprojectconfig/json_parser_test.go]: depends on what you want to benchmark, I chose to benchmark everything,  treat it as an example, it won't even run during automated tests

File: examples/README.md (hunk 3 above)
Comment (pawels-optimizely) [reply examples/README.md]: for one line, and just an example, I don't see a need.

File: examples/main_profile_feature.go (hunk 4 above)
Comment (pawels-optimizely) [reply examples/main_profile_feature.go]: does not matter, but I will remove it

File: examples/main_profile_feature.go (hunk 5 above)
Comment (pawels-optimizely) [reply examples/main_profile_feature.go]: does not matter, but I will remove it

File: examples/main_profile_feature.go (hunk 2 above)
Comment (pawels-optimizely) [reply examples/main_profile_feature.go]: ok, addressed

File: examples/main_profile_feature.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+	"github.com/optimizely/go-sdk/optimizely/client"
+	"github.com/optimizely/go-sdk/optimizely/decision"
+	"github.com/optimizely/go-sdk/optimizely/entities"
+	"github.com/optimizely/go-sdk/optimizely/notification"
+
+	"github.com/pkg/profile"
+)
+
+// stressTest has everything that test app has. it is used to run profile
```
[the remark this answers is not in the document]
Comment (pawels-optimizely) [reply examples/main_profile_feature.go]: ok, addressed

File: examples/main_profile_feature.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
+// to run the CPU profiling: go build -ldflags "-X main.RunCPUProfile=true" main.go && ./main
```
[the remark this answers is not in the document]
Comment (pawels-optimizely) [reply examples/main_profile_feature.go]: this is how you run profile: build it, run exe, run through profiling tool

File: examples/main_profile_feature.go [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+)
+
+// stressTest has everything that test app has. it is used to run profile
+func stressTest() {
+	/*
+		For the test app, the biggest json file is used with 100 entities.
+		DATAFILES_DIR has to be set to point to the path where 100_entities.json is located.
+	*/
+
+	var datafileDir = path.Join(os.Getenv("DATAFILES_DIR"), "100_entities.json")
```
[the remark this answers is not in the document]
Comment (pawels-optimizely) [reply examples/main_profile_feature.go]: ok, I will document it

File: optimizely/config/datafileprojectconfig/json_parser_test.go [hunk 10]
```diff
@@ -59,3 +62,18 @@ func TestParseDatafilePasses(t *testing.T) {
...
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		datafileStruct := &entities.Datafile{}
+
+		json.Unmarshal(datafile, &datafileStruct)
```
[the remark this answers is not in the document]
Comment (pawels-optimizely) [reply optimizely/config/datafileprojectconfig/json_parser_test.go]: ok, I will change it

File: examples/main_profile_feature.go [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+	clientApp.IsFeatureEnabled("feature_5", user)
+}
+
+var RunMemProfile = "false"
+var RunCPUProfile = "false"
+
+func main() {
+
+	if RunMemProfile == "true" || RunCPUProfile == "true" {
```
[the remark this answers is not in the document]
Comment (pawels-optimizely) [reply examples/main_profile_feature.go]: yeah, I can make it simpler

File: optimizely/config/datafileprojectconfig/json_parser_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (mikecdavis) [reply optimizely/config/datafileprojectconfig/json_parser_test.go]: yeah, but to be true to this package we should be benchmarking what this package is providing, which is just `Parse`, including the IO, while informative, will dilute the numbers generated.

File: examples/main_profile_feature.go [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+	clientApp.IsFeatureEnabled("feature_5", user)
+}
+
+var ProfileMode = ""
+
+const RUN_NUMBER = 50
+
+func main() {
+
+	if ProfileMode != "" {
```
Comment (mikecdavis) [inline examples/main_profile_feature.go]: is this check necessary, can't we handle this in the switch?
```
default:
   // log something bad
   return
}
...
```
as opposed to nesting.

File: examples/main_profile_feature.go (hunk 2 above)
Comment (mikecdavis) [reply examples/main_profile_feature.go]: still referencing "test app"

File: examples/README.md [hunk 13]
```diff
@@ -0,0 +1,9 @@
+# Optimizely Go SDK Examples
+
+### Profiling
+Prerequisite: `go get github.com/pkg/profile`
+
+DATAFILES_DIR env var has to be set to point to the path where 100_entities.json is located
```
Comment (mikecdavis) [inline examples/README.md]: Should probably add to the example code snippets below too.

File: examples/main_profile_feature.go (hunk 12 above)
Comment (pawels-optimizely) [reply examples/main_profile_feature.go]: done

File: examples/README.md (hunk 13 above)
Comment (pawels-optimizely) [reply examples/README.md]: add what ?

Comment (pawels-optimizely) [reply examples/README.md]: I added a quick description. i am sure there will be more, feel free to add more content if you like

**Pull request merged**